### PR TITLE
feat(core): increase coverage for data type formaters

### PR DESCRIPTION
**Pull request by ngouy**
## What is this about

add new type formatters

### CODE REVIEW

**Comment by ngouy** [comment]
@julienblin I will rebase the lodash things once your PR got merged

About the identifiers, I'm really not sure.
What's your take on it ?
https://github.com/bonfhir/bonfhir/pull/105/files#diff-9c2a583a7bd796347546fdfb9fbb475b3210ed4ba51d39fc16a39eea9718637bR61-R65

File: packages/core/r4b/data-types/address.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        dependentLocality: fhirAddress.district,
+        postalCode: fhirAddress.postalCode,
+        addressLines: fhirAddress.line,
+      });
+
+      switch (options?.style) {
+        case null:
+        case undefined:
+        case "text":
+          return fhirAddress.text || "";
```
Comment (julienblin) [inline packages/core/r4b/data-types/address.ts]: I think this should return "full" if `address.text` is missing. A fallback is a better solution IMHO.

File: packages/core/r4b/data-types/address.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import { removeDoubleSpaces } from "./helpers";
+import { fhirPeriodTypeAdapter } from "./period";
+
+/**
+ * An address expressed using postal conventions
+ *
+ * @see https://hl7.org/fhir/datatypes.html#address
+ */
+
+export type FhirAddressFormatOptions = {
```
Comment (julienblin) [inline packages/core/r4b/data-types/address.ts:16]: I believe we should include a `defaultCountry` option that would be used if the country is missing in the address data.

File: packages/core/r4b/data-types/address.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    // then sort by use
+    return (
+      addressUseOrder[address1.use || "undefined"] -
+      addressUseOrder[address2.use || "undefined"]
+    );
+  });
+};
+
+const dateTimeAdapter = fhirDateTimeTypeAdapter();
+const addressUseOrder = {
```
Comment (julienblin) [inline packages/core/r4b/data-types/address.ts]: This should be an option

File: packages/core/r4b/data-types/address.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+ * An address expressed using postal conventions
+ *
+ * @see https://hl7.org/fhir/datatypes.html#address
+ */
+
+export type FhirAddressFormatOptions = {
+  style?: "text" | "full" | "extended" | null | undefined;
+  useValueSetExpansions?: FhirCodeFormatOptions["valueSetExpansions"];
+  typeValueSetExpansions?: FhirCodeFormatOptions["valueSetExpansions"];
+};
```
Comment (julienblin) [inline packages/core/r4b/data-types/address.ts:67]: Also, when passing an array, we should be given the option to display a single address (first determine by the order), or all of them

File: packages/core/r4b/data-types/address.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        case "full":
+          return fullAddress.join("\n");
+        case "extended":
+          return removeDoubleSpaces(
+            _.compact([
+              `(${fhirPeriodTypeAdapter(locale).format(fhirAddress.period)})`,
+              `${type} - ${use}`,
+              ...fullAddress,
+              fhirAddress.country,
+            ]).join("\n")
```
Comment (julienblin) [inline packages/core/r4b/data-types/address.ts]: We should provide the option to format as a single line, or a paragraph

File: packages/core/r4b/data-types/address.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+  /**
+   * Format a FHIR address
+   *
+   * @see https://hl7.org/fhir/datatypes.html#address
+   */
+  format(
+    value: Address[] | Address | null | undefined,
+    options?: FhirAddressFormatOptions | null | undefined
+  ): string[] | string;
```
Comment (julienblin) [inline packages/core/r4b/data-types/address.ts]: I believe format should always return `string`, not `string[]`

File: packages/core/r4b/data-types/address.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  // JIT locale check
+  Intl.NumberFormat(locale);
+
+  return {
+    locale,
+    format(fhirAddress, options) {
+      if (!fhirAddress) return "";
+
+      if (Array.isArray(fhirAddress))
+        return sortAddresses(fhirAddress).map(
```
Comment (julienblin) [inline packages/core/r4b/data-types/address.ts]: This does not feel right. It format should always return a string IMHO.
We could leverage [Intl.ListFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat)

File: packages/core/r4b/data-types/address.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    const address1EndDate = dateTimeAdapter.parse(address1?.period?.end);
+    const address2EndDate = dateTimeAdapter.parse(address2?.period?.end);
+
+    // sort by period
+    if (!address1EndDate && address2EndDate) return -1;
+    if (!address2EndDate && address1EndDate) return 1;
+    if (address1EndDate && address2EndDate) {
+      if (address1EndDate.date > address2EndDate.date) return -1;
+      if (address2EndDate.date > address1EndDate.date) return 1;
+    }
```
Comment (julienblin) [inline packages/core/r4b/data-types/address.ts]: This feels like a `HasPeriodSorter` kinda helper

**Review by julienblin — CHANGES_REQUESTED** [review]
We need to review together the specs for those.

File: packages/core/r4b/data-types/address.ts (hunk 4 above)
Comment (ngouy) [reply packages/core/r4b/data-types/address.ts:67]: interesting
will do that

File: packages/core/r4b/data-types/address.ts (hunk 3 above)
Comment (ngouy) [reply packages/core/r4b/data-types/address.ts]: Option + this one as default?
Will do

**Review by julienblin — APPROVED** [review]

**Pull request merged**